File: mobile/lib/models/multimodal.dart
/// Multi-modal models for Voice (STT/TTS) and Image/Homework Scan integration.

enum AudioFormat {
  wav,
  mp3,
  aac,
  ogg,
  flac,
}

enum TTSPlayerState {
  stopped,
  buffering,
  playing,
  paused,
  completed,
  error,
}

enum ChildSafetyScanStatus {
  safe,
  flaggedNonEducational,
  flaggedPiiDetected,
  rejected,
}

class ChildVoiceParameters {
  final double pitchFrequencyHz;
  final double speedRate;
  final String ageGroup;

  ChildVoiceParameters({
    this.pitchFrequencyHz = 250.0,
    this.speedRate = 0.85,
    this.ageGroup = '6-7',
  });

  Map<String, dynamic> toJson() => {
        'pitch_frequency_hz': pitchFrequencyHz,
        'speed_rate': speedRate,
        'age_group': ageGroup,
      };

  factory ChildVoiceParameters.fromJson(Map<String, dynamic> json) {
    return ChildVoiceParameters(
      pitchFrequencyHz: (json['pitch_frequency_hz'] as num?)?.toDouble() ?? 250.0,
      speedRate: (json['speed_rate'] as num?)?.toDouble() ?? 0.85,
      ageGroup: json['age_group'] as String? ?? '6-7',
    );
  }
}

class STTAudioPayload {
  final String studentId;
  final String? audioBase64;
  final String? audioUrl;
  final String audioFormat;
  final int sampleRate;
  final String languageHint;
  final ChildVoiceParameters? childVoiceParams;

  STTAudioPayload({
    required this.studentId,
    this.audioBase64,
    this.audioUrl,
    this.audioFormat = 'wav',
    this.sampleRate = 16000,
    this.languageHint = 'en',
    this.childVoiceParams,
  });

  Map<String, dynamic> toJson() => {
        'student_id': studentId,
        'audio_base64': audioBase64,
        'audio_url': audioUrl,
        'audio_format': audioFormat,
        'sample_rate': sampleRate,
        'language_hint': languageHint,
        'child_voice_params': (childVoiceParams ?? ChildVoiceParameters()).toJson(),
      };
}

class STTResponse {
  final String transcribedText;
  final String detectedLanguage;
  final double confidence;
  final double durationSeconds;
  final bool piiStripped;
  final bool offlineFallback;

  STTResponse({
    required this.transcribedText,
    required this.detectedLanguage,
    required this.confidence,
    required this.durationSeconds,
    this.piiStripped = true,
    this.offlineFallback = false,
  });

  factory STTResponse.fromJson(Map<String, dynamic> json) {
    return STTResponse(
      transcribedText: json['transcribed_text'] as String,
      detectedLanguage: json['detected_language'] as String,
      confidence: (json['confidence'] as num).toDouble(),
      durationSeconds: (json['duration_seconds'] as num).toDouble(),
      piiStripped: json['pii_stripped'] as bool? ?? true,
      offlineFallback: json['offline_fallback'] as bool? ?? false,
    );
  }
}

class TTSRequest {
  final String studentId;
  final String text;
  final String language;
  final String voicePersona;
  final double speed;
  final double pitch;
  final String audioFormat;

  TTSRequest({
    required this.studentId,
    required this.text,
    this.language = 'en',
    this.voicePersona = 'suman_warm',
    this.speed = 0.85,
    this.pitch = 1.0,
    this.audioFormat = 'mp3',
  });

  Map<String, dynamic> toJson() => {
        'student_id': studentId,
        'text': text,
        'language': language,
        'voice_persona': voicePersona,
        'voice_params': {
          'speed': speed,
          'pitch': pitch,
          'volume': 1.0,
        },
        'audio_format': audioFormat,
      };
}

class WordTimestamp {
  final String word;
  final double startTime;
  final double endTime;

  WordTimestamp({
    required this.word,
    required this.startTime,
    required this.endTime,
  });

  factory WordTimestamp.fromJson(Map<String, dynamic> json) {
    return WordTimestamp(
      word: json['word'] as String,
      startTime: (json['start_time'] as num).toDouble(),
      endTime: (json['end_time'] as num).toDouble(),
    );
  }
}

class TTSResponse {
  final String audioBase64;
  final String audioFormat;
  final double durationSeconds;
  final int sampleRate;
  final List<WordTimestamp> wordTimestamps;
  final bool offlineFallback;

  TTSResponse({
    required this.audioBase64,
    required this.audioFormat,
    required this.durationSeconds,
    required this.sampleRate,
    required this.wordTimestamps,
    this.offlineFallback = false,
  });

  factory TTSResponse.fromJson(Map<String, dynamic> json) {
    return TTSResponse(
      audioBase64: json['audio_base64'] as String,
      audioFormat: json['audio_format'] as String,
      durationSeconds: (json['duration_seconds'] as num).toDouble(),
      sampleRate: json['sample_rate'] as int? ?? 22050,
      wordTimestamps: (json['word_timestamps'] as List<dynamic>?)
              ?.map((e) => WordTimestamp.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      offlineFallback: json['offline_fallback'] as bool? ?? false,
    );
  }
}

class BoundingBox {
  final double xMin;
  final double yMin;
  final double xMax;
  final double yMax;

  BoundingBox({
    required this.xMin,
    required this.yMin,
    required this.xMax,
    required this.yMax,
  });

  factory BoundingBox.fromJson(Map<String, dynamic> json) {
    return BoundingBox(
      xMin: (json['x_min'] as num).toDouble(),
      yMin: (json['y_min'] as num).toDouble(),
      xMax: (json['x_max'] as num).toDouble(),
      yMax: (json['y_max'] as num).toDouble(),
    );
  }
}

class OCRTextRegion {
  final String text;
  final double confidence;
  final BoundingBox boundingBox;
  final int lineNumber;

  OCRTextRegion({
    required this.text,
    required this.confidence,
    required this.boundingBox,
    required this.lineNumber,
  });

  factory OCRTextRegion.fromJson(Map<String, dynamic> json) {
    return OCRTextRegion(
      text: json['text'] as String,
      confidence: (json['confidence'] as num).toDouble(),
      boundingBox: BoundingBox.fromJson(json['bounding_box'] as Map<String, dynamic>),
      lineNumber: json['line_number'] as int,
    );
  }
}

class ImageScanPayload {
  final String studentId;
  final String? imageBase64;
  final String? imageUrl;
  final String mimeType;
  final String? subjectHint;

  ImageScanPayload({
    required this.studentId,
    this.imageBase64,
    this.imageUrl,
    this.mimeType = 'image/jpeg',
    this.subjectHint,
  });

  Map<String, dynamic> toJson() => {
        'student_id': studentId,
        'image_base64': imageBase64,
        'image_url': imageUrl,
        'mime_type': mimeType,
        'subject_hint': subjectHint,
      };
}

class ImageScanResult {
  final String scanId;
  final bool isEducational;
  final String safetyStatus;
  final String detectedText;
  final List<OCRTextRegion> textRegions;
  final String homeworkType;
  final List<String> childSafetyTags;
  final bool piiRedacted;
  final bool offlineFallback;

  ImageScanResult({
    required this.scanId,
    required this.isEducational,
    required this.safetyStatus,
    required this.detectedText,
    required this.textRegions,
    required this.homeworkType,
    required this.childSafetyTags,
    this.piiRedacted = true,
    this.offlineFallback = false,
  });

  factory ImageScanResult.fromJson(Map<String, dynamic> json) {
    return ImageScanResult(
      scanId: json['scan_id'] as String,
      isEducational: json['is_educational'] as bool,
      safetyStatus: json['safety_status'] as String,
      detectedText: json['detected_text'] as String,
      textRegions: (json['text_regions'] as List<dynamic>?)
              ?.map((e) => OCRTextRegion.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      homeworkType: json['homework_type'] as String? ?? 'worksheet',
      childSafetyTags: (json['child_safety_tags'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      piiRedacted: json['pii_redacted'] as bool? ?? true,
      offlineFallback: json['offline_fallback'] as bool? ?? false,
    );
  }
}

File: backend/app/schemas/multimodal.py
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field


class AudioFormat(str, Enum):
    WAV = "wav"
    MP3 = "mp3"
    AAC = "aac"
    OGG = "ogg"
    FLAC = "flac"


class ImageMimeType(str, Enum):
    JPEG = "image/jpeg"
    PNG = "image/png"
    WEBP = "image/webp"


class ChildSafetyScanStatus(str, Enum):
    SAFE = "safe"
    FLAGGED_NON_EDUCATIONAL = "flagged_non_educational"
    FLAGGED_PII_DETECTED = "flagged_pii_detected"
    REJECTED = "rejected"


class TTSVoicePersona(str, Enum):
    SUMAN_WARM = "suman_warm"
    SURESH_PLAYFUL = "suresh_playful"


class ChildVoiceParameters(BaseModel):
    pitch_frequency_hz: Optional[float] = Field(
        default=250.0,
        description="Average fundamental frequency (Hz) typical for child voices (200-350 Hz)",
    )
    speed_rate: float = Field(
        default=0.85,
        description="Speaking rate multiplier adjusted for primary children (0.7 - 1.2)",
    )
    age_group: str = Field(
        default="6-7",
        description="Target age group bracket (e.g. '5-6', '6-7')",
    )


class STTAudioPayload(BaseModel):
    student_id: str = Field(..., description="Unique student/user identifier")
    audio_base64: Optional[str] = Field(None, description="Base64 encoded audio binary data")
    audio_url: Optional[str] = Field(None, description="URL pointing to audio file")
    audio_format: AudioFormat = Field(default=AudioFormat.WAV, description="Audio container format")
    sample_rate: int = Field(default=16000, description="Sampling rate in Hz")
    language_hint: str = Field(default="en", description="Language hint ('en', 'hi', 'mr')")
    child_voice_params: Optional[ChildVoiceParameters] = Field(
        default_factory=ChildVoiceParameters,
        description="Child voice acoustic characteristics",
    )


class STTResponse(BaseModel):
    transcribed_text: str = Field(..., description="Speech-to-Text transcribed output text")
    detected_language: str = Field(..., description="Detected spoken language ('en', 'hi', 'mr')")
    confidence: float = Field(..., ge=0.0, le=1.0, description="Transcription confidence score")
    duration_seconds: float = Field(..., ge=0.0, description="Audio clip duration in seconds")
    pii_stripped: bool = Field(default=True, description="Indicates if PII/metadata was stripped")
    offline_fallback: bool = Field(
        default=False,
        description="True if transcription used local offline model fallback",
    )


class ChildTTSParameters(BaseModel):
    pitch: float = Field(
        default=1.0, description="Pitch modifier (0.8 - 1.2), calibrated for clarity"
    )
    speed: float = Field(
        default=0.85,
        description="Speaking speed (0.7 - 1.0) for primary school comprehension",
    )
    volume: float = Field(default=1.0, ge=0.0, le=1.0, description="Audio volume level")


class TTSRequest(BaseModel):
    student_id: str = Field(..., description="Unique student/user identifier")
    text: str = Field(
        ...,
        min_length=1,
        max_length=2000,
        description="Text string to synthesize into speech",
    )
    language: str = Field(default="en", description="Target language ('en', 'hi', 'mr')")
    voice_persona: TTSVoicePersona = Field(
        default=TTSVoicePersona.SUMAN_WARM, description="Voice persona profile"
    )
    voice_params: ChildTTSParameters = Field(
        default_factory=ChildTTSParameters,
        description="Audio speed/pitch parameters for child audience",
    )
    audio_format: AudioFormat = Field(
        default=AudioFormat.MP3, description="Requested output audio format"
    )


class WordTimestamp(BaseModel):
    word: str = Field(..., description="Individual spoken word")
    start_time: float = Field(..., description="Start offset in seconds")
    end_time: float = Field(..., description="End offset in seconds")


class TTSResponse(BaseModel):
    audio_base64: str = Field(..., description="Base64 encoded audio stream/payload")
    audio_format: AudioFormat = Field(..., description="Audio format of synthesized clip")
    duration_seconds: float = Field(
        ..., ge=0.0, description="Synthesized audio duration in seconds"
    )
    sample_rate: int = Field(default=22050, description="Audio sample rate in Hz")
    word_timestamps: List[WordTimestamp] = Field(
        default_factory=list,
        description="Word level alignment timestamps for synchronized highlighting",
    )
    offline_fallback: bool = Field(
        default=False, description="True if synthesis used local offline TTS engine"
    )


class BoundingBox(BaseModel):
    x_min: float = Field(
        ..., ge=0.0, le=1.0, description="Normalized minimum X coordinate (0.0 to 1.0)"
    )
    y_min: float = Field(
        ..., ge=0.0, le=1.0, description="Normalized minimum Y coordinate (0.0 to 1.0)"
    )
    x_max: float = Field(
        ..., ge=0.0, le=1.0, description="Normalized maximum X coordinate (0.0 to 1.0)"
    )
    y_max: float = Field(
        ..., ge=0.0, le=1.0, description="Normalized maximum Y coordinate (0.0 to 1.0)"
    )


class OCRTextRegion(BaseModel):
    text: str = Field(..., description="Extracted text from image region")
    confidence: float = Field(..., ge=0.0, le=1.0, description="OCR confidence score")
    bounding_box: BoundingBox = Field(..., description="Normalized bounding box coordinates")
    line_number: int = Field(..., ge=1, description="Line sequence number in worksheet")


class ImageResolution(BaseModel):
    width: int = Field(..., ge=1, le=4096, description="Image width in pixels")
    height: int = Field(..., ge=1, le=4096, description="Image height in pixels")


class ImageScanRequest(BaseModel):
    student_id: str = Field(..., description="Unique student/user identifier")
    image_base64: Optional[str] = Field(None, description="Base64 encoded image content")
    image_url: Optional[str] = Field(None, description="URL to worksheet image")
    mime_type: ImageMimeType = Field(default=ImageMimeType.JPEG, description="Image MIME type")
    resolution: Optional[ImageResolution] = Field(None, description="Image resolution dimensions")
    subject_hint: Optional[str] = Field(
        None, description="Optional subject context ('math', 'english', 'hindi', etc.)"
    )


class ImageScanResponse(BaseModel):
    scan_id: str = Field(..., description="Unique identifier for scan transaction")
    is_educational: bool = Field(
        ..., description="True if photo is validated as educational homework/worksheet"
    )
    safety_status: ChildSafetyScanStatus = Field(
        ..., description="Child safety scan assessment result"
    )
    detected_text: str = Field(..., description="Full combined OCR text extracted from worksheet")
    text_regions: List[OCRTextRegion] = Field(
        default_factory=list,
        description="Structured text blocks with bounding box positions",
    )
    homework_type: str = Field(
        default="worksheet",
        description="Identified type ('math_worksheet', 'handwriting_practice', etc.)",
    )
    child_safety_tags: List[str] = Field(
        default_factory=list, description="Safety tags assigned during safety inspection"
    )
    pii_redacted: bool = Field(
        default=True, description="True if EXIF/personal identifying metadata was stripped"
    )
    offline_fallback: bool = Field(
        default=False, description="True if offline mock/local vision model was utilized"
    )
